sql-plugin/src/main/331+/scala/com/nvidia/spark/rapids/GpuCheckOverflowInTableInsert.scala

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids
+
+import com.nvidia.spark.rapids.shims.ShimUnaryExpression
+
+import org.apache.spark.sql.rapids.shims.spark331plus.CastOverflowUtil
+import org.apache.spark.sql.types.DataType
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+/**
+ * Handle the new Expression added in Spark-3.3.1+.
+ * Casting a numeric value as another numeric type in store assignment. It captures the arithmetic
+ * exception thrown by Cast and shows a relevant error message.
+ * The current implementation in Spark creates a child of type Cast with AnsiEnabled set to true.
+ *
+ * The GPU equivalent columnar evaluation is a delegation to the child's columnar evaluation which
+ * is a GpuCast. In order to match Spark's Exception, we need to catch the Exception passing it to
+ * QueryExecutionError.
+ * The calculation of the sideEffect is delegated to the child Expression "GpuCast".
+ */
+case class GpuCheckOverflowInTableInsert(child: GpuCast, columnName: String)
+  extends ShimUnaryExpression with GpuExpression {
+
+  override def dataType: DataType = child.dataType
+
+  override def columnarEval(batch: ColumnarBatch): Any = {
+    try {
+      child.columnarEval(batch)
+    } catch {
+      // map SparkArithmeticException to SparkArithmeticException("CAST_OVERFLOW_IN_TABLE_INSERT")
+      case e if CastOverflowUtil.isCastOverflowException(e) =>
+        throw CastOverflowUtil.castCausingOverflowInTableInsertError(
+          child.child.dataType, child.dataType, columnName)
```

Comment:
CastOverflowUtil seems like an abstraction that isn't worth the boilerplate.  It's called in exactly one place (here), and there's even a comment in CastOverflowUtil stating it's for here.  Things get a lot simpler to read and the code more closely matches the Spark implementation if we don't overengineer this until we actually need the abstractions.  YAGNI applies here, IMO.
```suggestion
      case e SparkArithmeticException =>
        throw QueryExecutionErrors.castingCauseOverflowErrorInTableInsert(
          child.child.dataType,
          child.dataType,
          columnName)
```

Reply:
`QueryExecutionErrors` and `SparkArithmeticException` are both private to `Spark`. So, I followed the usual shim pattern of abstracting Spark private classes/methods.

Reply:
The simpler approach is to put GpuCheckOverflowInTableInsert into the same package CastOverflowUtil.  It's overall less code to do so given how trivial GpuCheckOverflowInTableInsert is.